firedrake/dmplex.pyx

```diff
@@ -1799,3 +1801,126 @@ def exchange_cell_orientations(
+        np.ndarray[PetscInt, ndim=1, mode="c"] val
+
+    val = np.empty(lsec.getStorageSize(), dtype=PETSc.IntType)
+    pStart, pEnd = lsec.getChart()
+
+    for p in range(pStart, pEnd):
+        CHKERR(PetscSectionGetDof(lsec.sec, p, &dof))
+        if dof > 0:
+            CHKERR(PetscSectionGetOffset(lsec.sec, p, &loff))
+            CHKERR(PetscSectionGetOffset(gsec.sec, p, &goff))
+            goff = cabs(goff)
+            for c in range(dof):
+                val[loff + c] = goff + c
+    return val
+
+
+def make_sf(dm, PETSc.Section lsec, PETSc.Section gsec):
+    """Make a PETSc SF for the provided local and global sections.
+
+    :arg dm: The DM to use to create the SF.
+    :arg lsec: Section describing the local data layout.
+    :arg gsec: Section describing the global data layout."""
+    cdef:
+        PETSc.SF sf1, sf2
+    dm.createDefaultSF(lsec, gsec)
+    sf1 = dm.getDefaultSF()
+    sf1.setUp()
+    sf2 = PETSc.SF()
+    # We need one of these per function space, and need to copy
+    # because we might change the default SF in the DM behind this
+    # object's back.
+    CHKERR(PetscSFDuplicate(sf1.sf, PETSCSF_DUPLICATE_GRAPH, &sf2.sf))
+    sf2.setFromOptions()
+    return sf2
+
+def halo_begin(PETSc.SF sf, dat, MPI.Datatype dtype, reverse):
+    """Begin a halo exchange.
+
+    :arg sf: the PETSc SF to use for exchanges
+    :arg dat: the :class:`pyop2.Dat` to perform the exchange on
+    :arg dtype: an MPI datatype describing the unit of data
+    :arg reverse: should a reverse (local-to-global) exchange be
+        performed.
+
+    Forward exchanges are implemented using :data:`PetscSFBcastBegin`,
+    reverse exchanges with :data:`PetscSFReduceBegin`.  Note that in
+    the latter case, the matching :func:`halo_end` call *must* be made
+    before the input data is modified.
+    """
+    cdef:
+        MPI.Op op = MPI.SUM
+        np.ndarray input, output
+
+    # Not allowed to touch input buffer (or observe output buffer)
+    # before the matching XXXEnd call is made.  To allow for
+    # overlapped computation (writing into local part) and
+    # communication (getting halo data) we therefore need a temporary
+    # buffer.
+    if not hasattr(dat, '_exchange_buf'):
+        dat._exchange_buf = np.empty_like(dat._data)
+    if reverse:
+        # Output goes to zeroed buffer, input buffer cannot be
+        # modified before matching halo_end call.
+        input = dat._data
+        output = dat._exchange_buf
+        output.fill(0)
+        CHKERR(PetscSFReduceBegin(sf.sf, dtype.ob_mpi,
+                                  <const void*>input.data,
+                                  <void *>output.data,
+                                  op.ob_mpi))
+    else:
+        # data copied to input buffer, output goes to input buffer,
+        # writes can occur to data in local region before matching
+        # halo_end.
+        input = dat._exchange_buf
+        np.copyto(input, dat._data, casting='no')
```

Comment:
Do we need to copy the entire `dat._data`, or can we get away with "owned" dofs only?